File: arxiv/anc/verify.py
#!/usr/bin/env python3
"""Standalone verifier for independent sets in strong powers of a cycle.

An independent verification path, written so that it shares no code with verify.mjs:
if both agree, a bug would have to be present in two separate implementations.

    python verify.py alpha-C7-6-1129.txt 7 6

A certificate line is one vertex of C_q^{boxtimes n}, given as n digits (first digit is
coordinate 0). Blank lines and lines starting with '#' are ignored. Two distinct vertices
are adjacent iff their cyclic distance is at most 1 in every coordinate, so the set is
independent iff every pair is at cyclic distance at least 2 in some coordinate.
"""
import sys


def cyclic_distance(a, b, q):
    d = abs(a - b) % q
    return min(d, q - d)


def main():
    if len(sys.argv) < 2:
        print("usage: python verify.py <certificate file> [q=7] [n=6]", file=sys.stderr)
        return 2
    path = sys.argv[1]
    q = int(sys.argv[2]) if len(sys.argv) > 2 else 7
    n = int(sys.argv[3]) if len(sys.argv) > 3 else 6

    vertices = []
    with open(path, encoding="utf-8") as handle:
        for lineno, raw in enumerate(handle, start=1):
            line = raw.strip()
            if not line or line.startswith("#"):
                continue
            parts = line.replace(",", " ").split() if (" " in line or "," in line) else list(line)
            digits = [int(p) for p in parts]
            if len(digits) != n:
                print(f"line {lineno}: expected {n} coordinates, got {len(digits)}", file=sys.stderr)
                return 1
            if any(d < 0 or d >= q for d in digits):
                print(f"line {lineno}: coordinate outside 0..{q - 1}", file=sys.stderr)
                return 1
            vertices.append(tuple(digits))

    if len(set(vertices)) != len(vertices):
        print("certificate contains duplicate vertices", file=sys.stderr)
        return 1

    violations = 0
    for i in range(len(vertices)):
        for j in range(i + 1, len(vertices)):
            if all(cyclic_distance(vertices[i][c], vertices[j][c], q) <= 1 for c in range(n)):
                violations += 1
                if violations <= 5:
                    a = "".join(map(str, vertices[i]))
                    b = "".join(map(str, vertices[j]))
                    print(f"adjacent pair: {a} and {b}", file=sys.stderr)

    if violations:
        print(f"NOT INDEPENDENT: {violations} adjacent pairs", file=sys.stderr)
        return 1

    size = len(vertices)
    print(f"OK: {size} distinct vertices of C_{q}^{n}, pairwise non-adjacent")
    print(f"    alpha(C_{q}^{n}) >= {size}, so Theta(C_{q}) >= {size}^(1/{n}) = {size ** (1.0 / n):.8f}")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: arxiv/anc/verify.mjs
// Standalone verifier for independent sets in strong powers of a cycle.
//
// No dependencies, no build step. Reads a certificate file whose lines are vertices of
// C_q^{boxtimes n}, written as n digits (the first digit is coordinate 0); blank lines and
// lines beginning with '#' are ignored.
//
//   node verify.mjs alpha-C7-6-1129.txt 7 6
//
// Two vertices of C_q^{boxtimes n} are ADJACENT iff they are distinct and their cyclic
// distance is at most 1 in EVERY coordinate; equivalently a set is independent iff every
// pair of distinct members is at cyclic distance at least 2 in at least one coordinate.
// So an independent set is exactly a code in Z_q^n with minimum cyclic Chebyshev distance 2.
//
// The check below is the naive O(|S|^2 n) all-pairs comparison, chosen deliberately: it is
// short enough to audit by eye, which matters more here than speed. For |S| = 1129 it runs
// in well under a second.

import { readFileSync } from 'node:fs';

const file = process.argv[2];
const q = Number(process.argv[3] ?? 7);
const n = Number(process.argv[4] ?? 6);
if (!file || !Number.isInteger(q) || q < 3 || !Number.isInteger(n) || n < 1) {
  console.error('usage: node verify.mjs <certificate file> [q=7] [n=6]');
  process.exit(2);
}

const vertices = [];
const lines = readFileSync(file, 'utf8').split(/\r?\n/);
for (let i = 0; i < lines.length; i++) {
  const line = lines[i].trim();
  if (line === '' || line.startsWith('#')) continue;
  const digits = /[,\s]/.test(line) ? line.split(/[,\s]+/).map(Number) : [...line].map(Number);
  if (digits.length !== n) { console.error(`line ${i + 1}: expected ${n} coordinates, got ${digits.length}`); process.exit(1); }
  for (const d of digits) {
    if (!Number.isInteger(d) || d < 0 || d >= q) { console.error(`line ${i + 1}: coordinate ${d} outside 0..${q - 1}`); process.exit(1); }
  }
  vertices.push(digits);
}

const cyclicDistance = (a, b) => { const d = Math.abs(a - b) % q; return Math.min(d, q - d); };

// duplicates
const seen = new Set();
for (let i = 0; i < vertices.length; i++) {
  const key = vertices[i].join(',');
  if (seen.has(key)) { console.error(`duplicate vertex ${key}`); process.exit(1); }
  seen.add(key);
}

// all pairs
let violations = 0;
for (let i = 0; i < vertices.length; i++) {
  for (let j = i + 1; j < vertices.length; j++) {
    let far = false;
    for (let c = 0; c < n; c++) {
      if (cyclicDistance(vertices[i][c], vertices[j][c]) >= 2) { far = true; break; }
    }
    if (!far) {
      violations++;
      if (violations <= 5) console.error(`adjacent pair: ${vertices[i].join('')} and ${vertices[j].join('')}`);
    }
  }
}

if (violations > 0) { console.error(`NOT INDEPENDENT: ${violations} adjacent pairs`); process.exit(1); }
console.log(`OK: ${vertices.length} distinct vertices of C_${q}^${n}, pairwise non-adjacent`);
console.log(`    alpha(C_${q}^${n}) >= ${vertices.length}, so Theta(C_${q}) >= ${vertices.length}^(1/${n}) = ${Math.pow(vertices.length, 1 / n).toFixed(8)}`);
